src/compiler/checker.ts

```diff
@@ -21363,7 +21360,139 @@ namespace ts {
+                     *           return x;  // A
+                     *       }
+                     *       x;  // B
+                     *       if ('B' === x.type) {  // second trigger
+                     *           return x;  // B
+                     *       }
+                     *       x;  // In the second time narrowing, the computed type would only have one(not Union), but should still be triggered to narrow type to never.
+                     *   }
+                     */
+                    // use declaredType is a quick way to resolve issue above -- when computed type is not union, the declared type is not chagned.
+                    // But it also bring some other issue, like #39114 and #39110. As long as the declared type is not proper, issues come.
+
+                    // and only declared union is too loose to judge it could be narrowed.
+                    /**
+                     * Condition 2: Index type with undefined/null
+                     *   function ffff1() {
+                     *       let a: Array<number>|undefined;
+                     *       a = [1];
+                     *       if(a){
+                     *           if (a[2] !== undefined) {
+                     *               a[1];
+                     *           } else {
+                     *               a[2];  // declared type is union, but here should not be narrowed to never.
+                     *           }
+                     *       }
+                     *   }
+                     */
+                    if (!(declaredType.flags & TypeFlags.Union)) {
+                        return false;
+                    }
+                    //, the types expect primitive should have more than one.
+                    if ((<UnionType>declaredType).types.filter(t => !(t.flags & TypeFlags.Primitive)).length < 2) {
+                        return false;
+                    }
+                    // However, I still use declared type, and put it here. The main aim is for future improving.
+                    // We could deal with 'this' type and use ugly code to deal with Condition 2, then we could avoid use declared type at least for now
+                    // passing all tests and issue #39110/#39114, but does it deserve? Is the code clear enough?
+                    return true;
+                }
+
+                // if root has undefiend|null, this need deal with differently.
+                let isRootHasUndefinedOrNull = false;
+                (<UnionType>computedType).types.forEach(t => {
+                    if (t.flags & TypeFlags.Nullable) {
+                        isRootHasUndefinedOrNull = true;
+                    }
+                });
+
+                return isTypeArrayDiscriminant(propertyTypeArray, isRootHasUndefinedOrNull);
+            }
+
+            function narrowTypeByDiscriminantNew(type: Type, access: AccessExpression, narrowTypeCb: (t: Type) => Type): Type {

```

Comment:
I add postfix because there is another function really named `narrowType`, I confuse them for some times.
I have no idea, is it a bad choice in fact?